feat: support for multiple TES versions

lbeckman314 opened the pull request:
# Overview

Resolves #48 (Add support for multiple TES versions in the test reports)

Happy to update or remove anything based on your discretion!

## Changes

### Multiple `--version` Flags

Allow for multiple `--version` flags to be passed to `cli.py`. Example of testing a TES server running on "http://localhost:8000/":

```sh
tes-compliance-suite report --version "1.0.0" --version "1.1.0" --server "http://localhost:8000/" --tag "get service info test" --serve
```

### Multiple Report Pages

Add a report page for every TES version passed in as above.

**TES v1.0 Report**

![tes-v1 0](https://user-images.githubusercontent.com/15309567/235263203-dbe1359e-6e0c-4312-8888-10ca9feb997c.png)

**TES v1.1 Report**

![tes-v1 1](https://user-images.githubusercontent.com/15309567/235263208-aeb50370-7aa9-43dc-b3a8-9dc2cde6130f.png)

LakiG committed:
Update TES v1.1 support

lbeckman314 committed:
Add support for multiple versions of TES reports

lbeckman314 committed:
Merge branch 'dev' into feature/tesv1.1

lbeckman314 committed:
Fix Flake8 linting error

lbeckman314 committed:
Minor formatting update

lbeckman314 commented:
Ah, working on the failing tests now!

lbeckman314 committed:
Update unittests to accept new TES versions

uniqueg commented on compliance_suite/cli.py:
```diff
@@ -20,14 +20,14 @@ def main() -> None:
 @main.command(help='Run TES compliance tests against the servers')
 @click.option('--server', '-s', help='server URL on which the compliance tests are run. Format - https://<url>/')
-@click.option('--version', '-v', help='TES version. Example - "1.0.0"')
+@click.option('--version', '-v', 'versions', multiple=True, help='TES version. Example - "1.0.0"')
```
Maybe change the description, too? E.g., "One or more TES versions. Example [...]"

uniqueg commented on compliance_suite/cli.py:
```diff
@@ -38,41 +38,55 @@ def report(server: str,
         output_path (str): The output path to store the JSON compliance report
         serve (bool): If true, runs a local server and displays the JSON report in webview
-        port (int): Set the local server port. Default - 16800
+        port (int): Set the local server port. Default - 15800
```
Any particular reason for the port change?

lbeckman314 replied:
Ah hopefully this reconciles the existing default port used by Click, happy to change it either way! 

```py
@click.option('--port', default=15800, help='port at which the compliance report is served')
```

https://github.com/elixir-cloud-aai/tes-compliance-suite/blob/dev/compliance_suite/cli.py#L27

uniqueg replied:
No, it's fine with me, if @LakiG doesn't have a particular reason to keep `16800`.

LakiG replied:
Thanks for pointing that out.

uniqueg commented on compliance_suite/cli.py:
```diff
@@ -38,41 +38,55 @@ def report(server: str,
     Args:
         server (str): The server URL on which the compliance suite will be run. Format - https://<url>/
-        version (str): The compliance suite will be run against this TES version. Example - "1.0.0"
+        version (List[str]): The compliance suite will be run against this TES version. Example - "1.0.0"
```
Again, would be good to change the description here as well so that it is clear that it can be a list of versions.

uniqueg commented on compliance_suite/cli.py:
```diff
@@ -38,41 +38,55 @@ def report(server: str,
+        if output_path is not None:
+            logger.info(f"Writing JSON Report on directory {output_path}")
+            with open(os.path.join(output_path, f"report-tes-{version}.json"), "w+") as output:
+                output.write(json_report)
```
It seems to me that in this way, _all_ available tests for a given version run for every version the user supplies. However, in many cases, tests will be applicable for multiple versions. And in these cases, the test results could be reported for multiple versions, too, to minimize runtime and save resources.

LakiG replied:
If a test is executed by any of the versions, we should not repeat it again as the result would be the same every time.
We'll need to explore how to store the state of executed tests.

Another thing to note, when generating the report, we'll need the copy and append the common test result of the first version to the other versions. Otherwise, they'll miss out on tests.

uniqueg commented on compliance_suite/report_server.py:
```diff
@@ -7,6 +7,7 @@
 import http.server
 import json
 import os
+from typing import List
```
Please follow import ordering used in library. Not sure how @LakiG deals with `from x import y` versus `import x`, but it seems to me that at least within each of these, there is alphabetical sorting according to the package name.

LakiG replied:
Yeah, there doesnt seem to be any guideline about this. Ref - https://peps.python.org/pep-0008/#imports
But within the same import group, it would be better to separate them. Example:

`import x` _All alphabetically sorted_
`from x import y` _All alphabetically sorted_

uniqueg commented on compliance_suite/report_server.py:
```diff
@@ -64,7 +67,7 @@ def start_local_server(
         self.local_server = socketserver.TCPServer(("", port), http_handler)
         logger.info(f"Starting a local server at  http://localhost:{port}")
-        webbrowser.open(f"http://localhost:{port}")
+        webbrowser.open(f"http://localhost:{port}/report-tes-{self.versions[0]}.html")
```
Is this how the default version to show is set? If so, I'm not sure the behavior is as you say, i.e., that the version the user specifies _first_ is the one that is shown. Rather, it would, I think, be always the smallest version, because of `versions.sort()` in the CLI module.

However, personally I'm actually not quite sure there's a lot of benefit to allow the user specifying the order of the reports (and if so, that behavior should be documented somewhere). Rather, I would always just start with the latest version.

uniqueg commented on compliance_suite/web/index.html:
Was this deliberately deleted? Is it maybe referred to in the docs, as an example? If it was deleted because it was inadvertently added, perhaps it should also be added to `.gitignore`?

LakiG replied:
Although its not being referred, it could serve as an example file.

We can refer to this in the docs as well when we have developer docs specifically.

uniqueg commented on compliance_suite/web/partials/navbar.html:
```diff
@@ -1,3 +1,6 @@
 <nav id="navbar" class="navbar navbar-expand-lg navbar-light bg-light">
-    <a id="navbar-title" class="navbar-brand" href="#">GA4GH TES v0.1.0 Compliance Report</a>
+    <a id="navbar-title" class="navbar-brand" href="#">GA4GH TES Compliance Report</a>
+    {% for version in versions %}
+        <a id="navbar-title" class="navbar-brand" href="report-tes-{{version}}.html">TES v{{version}}</a>
+    {% endfor %}
```
I think this is good opportunity to make the test runner more generic by not only not hardcoding the version, but also not hardcoding the API _name_ (here: "GA4GH TES"). Also, "Compliance" is perhaps too specific, I would rather go with "Test Report".

LakiG replied:
I agree, @lbeckman314 added a new field `name`. That can be used here.

uniqueg commented on compliance_suite/web/views/report.html:
```diff
@@ -1,9 +1,8 @@
-
+{% extends "views/base.html" %}
+{% block report %}
 <div class="container">
 
-  <h2>TES Compliance Report</h2>
+  <h2>TES v{{version}} Compliance Report</h2>
```
Same as above: If not too difficult, try to get rid of "TES" as well.

uniqueg commented on compliance_suite/web/web_report.json:
Same as above: just checking whether deletion was deliberate and whether it may make sense to add this file to `.gitignore`

uniqueg commented on tests/create_task_streamable.yml:
```diff
@@ -26,7 +26,7 @@ jobs:
           }
         ],
-        "input": [
+        "inputs": [
```
I guess this is technically a part of another issue.

uniqueg commented on tests/create_task_output.yml:
Is this a new test? If so, maybe it should go to a different PR as it has nothing to do with the versions?

Also, is this really a test specific to TES 1.1.0? It seems to me that it should also be applicable to TES 1.0.0, but I may be wrong.

uniqueg reviewed (changes requested):
Good stuff! A couple of comments, some just to clarify things I'm not sure I understand.

There is, I think, also another issue that may be worth thinking about: If (and only if!) we can assume that the test runner will only be used with semantically versioned APIs, then I think supplying multiple minor or patch versions is pointless, because it will be expected that _every_, say, `1.0.0` test will be expected to yield the same result for `1.0.0` and `1.x.y` implementations. The same, btw, goes for the version tags. Rather than tagging them with multiple minor versions, tagging them with multiple major versions should be enough.

And while I think that it is probably too strong an assumption that all tested APIs will be semantically versioned, this assumption may already be hardcoded in the current test runner. So I see three options here:
1. Clearly state that assumption and update the test runner accordingly, for the version labeling of tests, i.e., only supply the minimum version for each _major_ version release of the API, e.g., `1.1.1` and `2.0.1`, for the selection of which tests to run (all tests with the same major and equal or lower minor/patch version combination), and for this PR, i.e., only keep the highest minor version for each major versions and log a warning if the user supplies any redundant minor version, e.g., if they supply `1.0.1`, `1.1.1`, `2.0.0` and `2.0.1`, then tests would only be run for `1.1.1` and `2.0.1`.
2. Refactor/modify all code that assumes that an API versions follows SemVer.
3. Allow the user to indicate whether an API follows SemVer or not (default: `True`), and depending on that information, either make use of the implications of SemVer adoption to to minimize test runs, or not.

Personally I tend to favor the last two options (potentially supporting more APIs). Ideally the last one (best of both worlds), but it may add too much complexity. What do you both think?

lbeckman314 committed:
Add 'name' option for API name in report

- Restore example report and data files

lbeckman314 committed:
Update report server tests to use new 'name' constant

lbeckman314 committed:
Add pytest fixtures for report server testing

lbeckman314 committed:
Remove extraneous unittest files

lbeckman314 committed:
Add unittest files with correct case of 'TES'

lbeckman314 committed:
Re-add web_report.json (mistaken deletion)

lbeckman314 committed:
Restore web_report.json

lbeckman314 committed:
Update web_report.json

lbeckman314 committed:
Update web_report.json

lbeckman314 committed:
Revert web_report.json changes

lbeckman314 committed:
Delete web_report.json

lbeckman314 committed:
Add test for exit code 1 (any tests failed)

lbeckman314 committed:
Small flake8 fix

lbeckman314 committed:
Update index.html behavior

- Slightly hack-ish, copies the latest schema HTML page to index.html, which is then opened by the report server.

lbeckman314 committed:
Add date to test report (UTC timezone, ISO format)

- e.g. Created: 2023-05-11T22:16:53 UTC

uniqueg commented:
@LakiG: Did you review the PR?
@lbeckman314: Will you tag me/us for re-review when you are done with changes from your side?

LakiG commented on compliance_suite/cli.py:
```diff
@@ -49,30 +52,48 @@ def report(server: str,
-            output.write(json_report)
+    logger.info(f"Input name - {name}")
+
+    any_fail = False
```
Please add the variable annotations for new variables.

LakiG commented on compliance_suite/cli.py:
```diff
@@ -49,30 +52,48 @@ def report(server: str,
+
+    # If report has failed tests, exit with error code
+    if any_fail is True:
+        exit(1)
```
In a scenario where `serve` is `True` and there were failed cases, shouldn't we get `exit(1)`?

LakiG commented on compliance_suite/job_runner.py:
```diff
@@ -55,6 +55,9 @@ def __init__(self, server: str, version: str, tags: List[str]):
         self.report: Any = None
 
+    def get_test_status(self) -> Dict[str, List[str]]:
+        return self.test_status
```
Maybe we can add the docstring for this new method

LakiG commented on compliance_suite/report_server.py:
```diff
@@ -4,24 +4,30 @@
 import jinja2 as j2
 
 from compliance_suite.functions.log import logger
+from pathlib import Path
```
We can move this to the first import group as it is a standard in-built library.

LakiG commented on compliance_suite/report_server.py:
```diff
@@ -31,22 +37,31 @@ def __init__(self, web_dir: str):
         self.local_server: Any = None
         self.web_dir: str = web_dir
+        self.versions: List[str] = versions
+        self.name = name
```
The variable notation is needed here as well.

LakiG commented on unittests/test_report_server.py:
```diff
@@ -4,24 +4,25 @@
     MagicMock,
     patch
 )
-
```
This line break is required to separate the import groups.

LakiG reviewed (changes requested):
You can modify the PR name since its not only applicable to test reports. We are adding support for multiple versions overall.

LakiG commented:
> Personally I tend to favor the last two options (potentially supporting more APIs). Ideally the last one (best of both worlds), but it may add too much complexity. What do you both think?

Currently, we are following the second point where we assume all API versions maintain SemVer. Though we earlier had a discussion on this, regarding the base URL to be used for non-SemVer APIs and couldn't reach a conclusion. We can add support for those if we face any such API later.

Regarding the point that a major version would give the same test results, it depends on how the API developers have maintained their versions. For example, if we see TES only, some new features were added in the minor version 1.1.0. The new tests would fail for the previous minor version 1.0.0, so essentially the same major version is giving different results.
 
cc - @uniqueg @lbeckman314

lbeckman314 committed:
Include response code in test runner output

lbeckman314 committed:
Add polling to create task ignore error test

- Updated polling increments from 10 seconds to 1 second

lbeckman314 committed:
Revert changes to example web report page (index.html)

lbeckman314 commented:
Apologies for the slow updates -- we've been using this branch of tes-compliance-suite to update[ Funnel's v1.0 and v1.1 compliance status](https://ohsu-comp-bio.github.io/funnel-compliance/report-TES-1.1.0.html) in preparation for it's newest release.

I'll plan on rebasing onto the latest updates in [elixir-cloud-aai:dev](https://github.com/elixir-cloud-aai/tes-compliance-suite/tree/dev) today to make any potential merges or recommendations easier!

Also happy to close this PR if development is using a different strategy for managing different TES versions.

Just a side note but we were able to briefly show off your all's work on the tes-compliance-suite this morning on the [GA4GH Cloud Work Stream](https://docs.google.com/document/d/1hayvWLIoymomPI9oXcaTZirn5YxFv1cYAs70zyvlvnA/edit#heading=h.o1ijhfdu9zv) meeting!

uniqueg commented:
Hi @lbeckman314 - thanks for the update! We'd be happy to merge this PR upstream once you have updated/rebased it, so please let us know when that is :)

Also, thanks for demonstrating the compliance suite at the GA4GH meeting :rocket:

vsmalladi commented:
@lbeckman314 @LakiG  can we merge this in?

lbeckman314 commented:
Ah, looking over this PR it might be best to close it and open separate ones for the following features:
- Report publishing for multiple versions
- Exit code 1 on test failure (for Github Action status reporting)

I will go ahead and close this PR and create these new more specific ones for review!
